### ISSUE TITLE
NodeGraph: Allow specifying fixed X and Y coordinates for nodes in DataFrame

### ISSUE THREAD

**Issue by timo** [grafana/grafana#86527]
**Why is this needed**:

I want to visualize data and statistics for nodes in an existing graph from another application where the user directly places nodes and their connections on a canvas. As such, the nodes I query from the database (postgresql) have X and Y coordinates available, and in order for the user to more easily recognize which nodes from that editor correspond to which nodes in grafana, it would be great to just take the positions from a dataframe. The layout algorithm makes rather a mess of things :)

**What would you like to be added**:

Fields in the DataFrame that would be picked up before the layouting algorithm runs. They could be named x/y, or fixedX/fixedY or staticX/staticY or something similar.

**Who is this feature for?**

Anyone who takes node graph data from a source that has positions already assigned, and anyone who wants to do their own layouting in the query (don't laugh, you can do incredible things in sql ;) )

**Additional notes**

I actually already have a "minimum viable patch" for this, but it would 1) need to be cleaned up and 2) I would like to find out first, if this is even something the grafana team would accept into upstream.

Without coordinates from data:
![image](https://github.com/grafana/grafana/assets/25167/2aca42b6-64cd-4451-a225-3b7b2c792cc3)
With coordinates from data:
![image](https://github.com/grafana/grafana/assets/25167/4bb2f84e-1e68-49c9-adbc-01c356b8969d)

You can see that the layout used here is not one that the layouter would ever come up with, what with nodes overlapping and everything (mostly just needs a `y * 2` in the sql query), but the ability to recognize which node is which is very valuable for this use case!

**Comment by aocenas** [comment]
Yeah I think it would make sense. I am bit warry of adding too much stuff into the data frame mainly cosmetics like colors, sizes etc as I am still not sure the dataFrame columns is the best way to represent it but coordinates are probably more meaningful for this.

I wonder though how it should behave in some edge cases. Like we have to skip our own layout in case we have coordinates, but we would have to check if all nodes have those I guess or maybe just render them at 0,0 or something. Also wonder if we should do anything about accidental overlaps for debugging purposes.

**Comment by timo** [comment]
The layout algorithm already supports fixed positions for nodes by not moving them around, but still allowing them to apply forces to other nodes, I think. So any nodes that are not fixed would still be moved into positions that make as much sense as possible given the constraints.

By saying these things shouldn't go into dataframe columns, would you suggest the dataframe columns have more "semantic" data, like "num_cpu_cores" for a server for example, and then somewhere else have a function outside of the data source that maps num_cpu_cores to size_of_node?

**Comment by aocenas** [comment]
> The layout algorithm already supports fixed positions for nodes by not moving them around, but still allowing them to apply forces to other nodes, I think.

We have 2 different layout algorithms now, one is D3 force based and second behind a feature flag is something closer to GraphViz layered algo that is more rule based that we want to move to. Not sure what that does if it works with predefined positions or not right now.

> By saying these things shouldn't go into dataframe columns

I guess this is more of a philosophical question right now. The dataFrame itself is for data, not sure if color of the node counts as "data". We have some other parts of the data frame that can specify some formatting options like `field.config` where we can say what unit a column has or some more specific stuff in `field.config.custom` for visualization to read and decide on some formatting options. So there are different ways we do formatting without embedding it directly into the data but there is probably nothing short-term we will change and the node graph has some specifics, like the need to format per row not per column.

**Comment by timo** [comment]
true, the per-row needs change the equation. though i would be amused by a time series chart where every point has a different color without having to make one series per point :)

it seems like i might have lost the changes that make a fixed x/y coordinate work. i'll look around a bit more, and then i can also check to see how the different algorithm behaves when given fixed x and/or y coordinates. there'll have to be a bunch of tests in the end anyway to ensure this does what we expect & want.

**Comment by timo** [comment]
Unfortunately I can't find anything in msagljs that looks like a way to pin individual nodes to a given position. Sounds like with the layout engines from that package, the only choices would be:

* give every node a fixed position, and skip the layouting algorithm entirely
* get positions for every node dictated by the layout engine
* take the nodes with fixed position out of the graph along with their incoming and outgoing edges, run the layouting algorithm, then put the nodes back with their fixed positions, or
* leave the nodes in the graph and just overwrite their positions with the fixed positions at the end, and accept that these nodes will most likely end up in pretty much random positions compared to the rest of the graph, with edges probably intersecting with about a hundred random nodes each

for my use case, every node would have a position assigned, so it wouldn't be that bad to be forced to choose layout engine entirely off or entirely on, but I don't know about other use cases.

you want to completely remove the force-based layouting algorithm from the codebase?

**Comment by aocenas** [comment]
Yeah I don't think mixing layout engine position with custom position is meaningful. Like as you say it would end up in some random positioning, that can also change if we do any tweak to the layout so I would not like people to count on it even if for some cases it may work.

So yeah if we do this I think it should be fully managed by the data explicitly or automatic layout and nothing in between and we can probably have some warning if some nodes don't have coordinates.

> you want to completely remove the force-based layouting algorithm from the codebase?

We are still sort of testing this. Right now there is a cutoff for some number of nodes where we switch to force based layout because of performance limitations. So the force based layout will be still there. The question is if there is a usecase for the force layout. Most our usecases are in the context of computing infrastructure like service graphs which look more reasonable with the layered based algorithm as they are more neatly "organised". For some more random graphs, the force layout may still be better though but it's not easy to automatically detect. There is a possibility of making it configurable in panel options at least for dashboards.

**Comment by timo** [comment]
When you say "at least for dashboards", do you mean as opposed to the explore view? Or is it maybe related to that new Scenes feature that I've been meaning to check out ...

**Comment by timo** [comment]
the mentioned pull request has my first stab at making this work. it comes with a screenshot and a ready-to-test shared dashboard.

I'd love your feedback on whether the data flows correctly (where "fixedX" gets translated to "x", how the EdgeDatum gets its source/target updated when layout() isn't used, stuff like that), and maybe if you have a wish for the kinds of test cases I should write; if I should focus on covering as many branches as I can for example.

**Comment by aocenas** [comment]
> When you say "at least for dashboards", do you mean as opposed to the explore view?

Yes exactly. We have the panel options in dashboards where you can customize the visualizations, in Explore though the idea is to just get the right visualization without user to have to tinker much with it so we don't allow much in sense of changing the configuration (although as counter example we have a bar/line/stack options for time series graph in explore)

Let's continue the second comment in the PR.
